rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java

```diff
@@ -454,21 +485,28 @@ private void collectAppl(ITree tree, @Nullable String parentCategory) {
 
             // Collect tokens from children
             for (IValue child : TreeAdapter.getArgs(tree)) {
-                //Propagate current category to child unless currently in a syntax nonterminal
-                //*AND* the current child is a syntax nonterminal too
-                if (!TreeAdapter.isChar((ITree) child) && ProductionAdapter.isSort(prod) &&
-                        ProductionAdapter.isSort(TreeAdapter.getProduction((ITree) child))) {
-                    collect((ITree) child, null);
-                } else {
-                    collect((ITree) child, category);
-                }
+
+                // Special case (syntax-in-syntax): In legacy mode, do *not*
+                // propagate `category` when both `tree` and `child` are syntax
+                // non-terminals.
+                var specialCase =
```

Comment:
There is a `isContextFree` method in `ProductionAdapter`, maybe this condition can be written even more clearly when using that?

Reply:
Using `isContextFree` would slightly change the semantics of the condition. As we're keeping the special case only for backward-compatibility at this point (it is planned to be removed in a future release), I think we need to keep the current formulation.